File: frontend/src/routes/-file/markers.ts
// TODO(ai-review): review for style and correctness
//
// Wire-format markers shared with the backend's `unity::markers`
// module. The backend emits each marker as a JSON string
//
//   __MARK__<type>␞<payload>
//
// inside the pretty-printed object dump. One regex on the post-shiki
// HTML matches every marker, then `renderMarker` dispatches on the
// type tag.
//
// Adding a new marker means three places: a type tag + format helper
// on the backend, an arm in the dump walker, and a matching case in
// `renderMarker` below.
//
// Currently shipped markers:
//
// - `pptr` — `__MARK__pptr␞<ref>␞<target>␞<type>␞<file>` (Unity PPtr).
//   Inside a structured-diff content body both sides emit byte-equal
//   markers — the per-side routing is taken from the unified-diff
//   `+`/`-` gutter, not baked into the marker shape.
// - `classref` — `__MARK__classref␞<ref>␞<name>␞<file>` (a managed
//   class reference, e.g. a MonoScript's `m_ClassName` linking into the
//   decompiled `Managed/<Assembly>.dll`; `<ref>` is a `type:<FQN>` hash
//   the dll structured-view jumps to). Always an external link, so it
//   reuses the pptr renderer's file→`?path=` / ref→`#hash` plumbing but
//   carries no pptr `(type)` suffix.
// - `color` — `__MARK__color␞#rrggbbaa` (rgba color).

import type { FileLocator } from "./types";

const MARK_PREFIX = "__MARK__";
const MARK_SEP = "␞";
// `[^"]*` for the payload is enough because JSON strings can't contain
// raw `"` (it'd close the string) and `␞` is JSON-stringify-stable —
// shiki keeps the whole marker inside one string token.
const MARKER_RE = new RegExp(`"${MARK_PREFIX}([a-z]+)${MARK_SEP}([^"]*)"`, "g");

/// Compose a post-shiki HTML rewrite that swaps each marker for its
/// rendered HTML. Unknown types pass through untouched so a backend
/// that ships a marker before the frontend understands it doesn't
/// crash the page.
///
/// `isLocalRefInTree` decides whether a same-file pptr ref is a real
/// jump target (rendered as a link) or a dead one (rendered as plain
/// text). The backend filters most dead refs (self-transforms, etc)
/// but some — like Transform refs on components when the tree only
/// lists GameObjects — still come through with the marker shape.
export function makePostProcess(
  locator: FileLocator,
  isLocalRefInTree: (ref: string) => boolean,
): (html: string) => string {
  const renderers = makeRenderers(locator, isLocalRefInTree);
  return (html) => applyMarkers(html, renderers);
}

/// Diff variant: pptr links on a removed (`-`) line resolve against the
/// `target` manifest, everything else (added `+`, context, headers)
/// against `base`. Both diff sides emit byte-equal markers — an
/// unchanged ref must not show up as a change — so the side isn't in
/// the marker; we recover it per line from the unified-diff gutter.
/// Handles both shiki HTML (markup stripped to find the gutter) and the
/// plain pre-shiki fallback.
export function makeDiffPostProcess(
  base: FileLocator,
  target: FileLocator,
  isLocalRefInTree: (ref: string) => boolean,
): (html: string) => string {
  const renderBase = makeRenderers(base, isLocalRefInTree);
  const renderTarget = makeRenderers(target, isLocalRefInTree);
  return (html) =>
    html
      .split("\n")
      .map((line) => applyMarkers(line, isRemovalLine(line) ? renderTarget : renderBase))
      .join("\n");
}

/// The per-locator renderers a marker can dispatch to. Bundled so both
/// the single-side and diff post-processors thread the same set through
/// `applyMarkers`. `color` needs no locator so it stays a free function.
type Renderers = {
  pptr: (payload: string) => string;
  classref: (payload: string) => string;
};

function makeRenderers(
  locator: FileLocator,
  isLocalRefInTree: (ref: string) => boolean,
): Renderers {
  return {
    pptr: makePptrRenderer(locator, isLocalRefInTree),
    classref: makeClassrefRenderer(locator),
  };
}

function applyMarkers(html: string, renderers: Renderers): string {
  return html.replace(MARKER_RE, (whole, type, payload) => {
    switch (type) {
      case "pptr":
        return renderers.pptr(payload);
      case "classref":
        return renderers.classref(payload);
      case "color":
        return renderColor(payload);
      default:
        return whole;
    }
  });
}

/// A rendered diff line is a removal when its first visible character —
/// past any markup shiki opened the line with — is `-`. Only the
/// leading tags are skipped, so this stays cheap on long, span-heavy
/// lines instead of stripping the whole line to read one char.
const REMOVAL_LINE_RE = /^(?:<[^>]*>)*-/;
function isRemovalLine(lineHtml: string): boolean {
  return REMOVAL_LINE_RE.test(lineHtml);
}

/// Build the `…/file?path=<depotPath>` href (plus the non-public
/// `&branch=` param) for one locator. Shared by the pptr and classref
/// renderers — both link to another depot file by path.
function makeFileHref(locator: FileLocator): (depotPath: string) => string {
  const branchParam =
    locator.branch === "public" ? "" : `&branch=${encodeURIComponent(locator.branch)}`;
  return (depotPath) =>
    `/apps/${locator.appid}/depots/${locator.depotId}/manifests/${locator.manifestId}/file?path=${encodeURIComponent(depotPath)}${branchParam}`;
}

/// A `classref` link: always external (it points into a decompiled
/// assembly), so it goes straight through the file-href + `#type:<FQN>`
/// hash plumbing. The click handler keys off `data-pptr-file="1"` to
/// route external links through tanstack-router, same as an external
/// pptr — no separate handler needed.
function makeClassrefRenderer(locator: FileLocator): (payload: string) => string {
  const fileHref = makeFileHref(locator);
  return (payload: string): string => {
    const [ref = "", name = "", file = ""] = payload.split(MARK_SEP);
    const label = escHTML(name);
    if (!ref || !file) {
      return `<span class="text-slate-400">${label}</span>`;
    }
    const href = `${escHTML(fileHref(file))}#${escHTML(ref)}`;
    return `<a href="${href}" data-pptr-ref="${escHTML(ref)}" data-pptr-file="1" class="cursor-pointer text-sky-400 underline decoration-sky-700 hover:decoration-sky-400 hover:text-sky-200">${label}</a>`;
  };
}

function makePptrRenderer(locator: FileLocator, isLocalRefInTree: (ref: string) => boolean) {
  const fileHref = makeFileHref(locator);
  // Only show the file's basename in the link label — the full depot
  // path (the `<Game>_Data/StreamingAssets/aa/StandaloneWindows64/…`
  // mouthful for addressables bundles) belongs in the `href` hover,
  // not on screen for every pptr in the dump.
  const shortFileLabel = (depotPath: string) => {
    const slash = depotPath.lastIndexOf("/");
    return slash >= 0 ? depotPath.slice(slash + 1) : depotPath;
  };
  return (payload: string): string => {
    const [ref = "", target = "", type = "", file = ""] = payload.split(MARK_SEP);
    // Fully-empty payload = null pptr that landed in a map-key
    // position. Render as plain `null` to avoid a stray "()".
    if (!ref && !target && !type && !file) {
      return '<span class="text-slate-500">null</span>';
    }
    // Label shape:
    //  - local refs: just the target name (the row in the tree they
    //    point at carries everything else).
    //  - external refs: `<file>: <name>` when the backend resolved a
    //    real `m_Name`; `<file> #<pathid>` otherwise. Either way the
    //    file is what makes the reference unambiguous.
    const ty = escHTML(type);
    const pathIdHint = (() => {
      const m = /^obj:(\d+)$/.exec(ref);
      return m ? ` <span class="text-slate-500">#${m[1]}</span>` : "";
    })();
    const hasResolvedName = target !== "" && !/^PathID=\d+$/.test(target);
    let label: string;
    let pathHint = "";
    if (file) {
      const shortFile = escHTML(shortFileLabel(file));
      if (hasResolvedName) {
        label = `${shortFile}<span class="text-slate-500">:</span> ${escHTML(target)}`;
      } else {
        label = shortFile;
        pathHint = pathIdHint;
      }
    } else {
      label = escHTML(target) || '<span class="text-slate-500">null</span>';
    }
    const suffix = `${ty ? ` <span class="text-slate-500">(${ty})</span>` : ""}${pathHint}`;
    if (!ref) {
      return `<span class="text-slate-400">${label}</span>${suffix}`;
    }
    // Same-file refs to objects the tree doesn't list (e.g. Transform
    // components when the tree only carries GameObjects) have no jump
    // target — render plain so the user doesn't chase a dead click.
    if (!file && !isLocalRefInTree(ref)) {
      return `<span class="text-slate-400">${label}</span>${suffix}`;
    }
    // Both local and external refs get a real `href` so browsers show
    // the target on hover and middle-/ctrl-click opens a new tab. The
    // click handler tells them apart via `data-pptr-file`: same-file
    // refs go through the in-page hash logic, external refs through
    // tanstack-router.
    const linkAttrs = file
      ? `href="${escHTML(fileHref(file))}#${escHTML(ref)}" data-pptr-ref="${escHTML(ref)}" data-pptr-file="1"`
      : `href="#${escHTML(ref)}" data-pptr-ref="${escHTML(ref)}"`;
    return `<a ${linkAttrs} class="cursor-pointer text-sky-400 underline decoration-sky-700 hover:decoration-sky-400 hover:text-sky-200">${label}</a>${suffix}`;
  };
}

function renderColor(payload: string): string {
  // The backend constrains the payload to `#` + 8 lower-case hex
  // chars; bail out on anything else so a malformed marker can't
  // smuggle attribute-breaking characters into the inline style.
  if (!/^#[0-9a-f]{8}$/.test(payload)) {
    return `<span class="text-slate-400">${escHTML(payload)}</span>`;
  }
  return `<span class="inline-block h-[0.9em] w-[0.9em] mr-1 rounded-sm border border-slate-700/60" style="background-color: ${payload}; vertical-align: -0.08em"></span><span class="text-slate-300">${payload}</span>`;
}

export function escHTML(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

File: crates/transform/src/unity/serializedfile/markers.rs
// TODO(ai-review): review for style and correctness
//! Wire-format markers shared between the JSON dump (backend) and the
//! preview HTML rewriter (frontend `markers.ts`).
//!
//! Every marker rides through the dump as a JSON *string* of the form
//!
//! ```text
//! __MARK__<type>␞<payload>
//! ```
//!
//! `␞` (U+241E) is a printable codepoint JSON.stringify passes through
//! unescaped, so shiki sees the whole sentinel as one string token —
//! one regex on the frontend reliably picks it back up. `<type>` is a
//! short lower-case tag the frontend switches on; `<payload>` is
//! marker-specific (can itself contain `␞`-separated fields).
//!
//! Adding a new marker means three places:
//! - new type tag + format helper here,
//! - one arm in the [`super::dump_value::simplify_for_dump`] walker,
//! - matching case in `markers.ts`'s `renderMarker`.
//!
//! Currently shipped markers:
//!
//! - `pptr` — `__MARK__pptr␞<ref>␞<target>␞<type>␞<file>` (Unity PPtr)
//! - `color` — `__MARK__color␞#rrggbbaa` (rgba color)

use std::collections::BTreeMap;
use std::fmt::Write as _;

use rabex_env::rabex::objects::pptr::{FileId, PPtr, PathId};
use serde_value::Value;

/// Prefix every marker starts with. Picked to be conspicuous in a JSON
/// dump (no real Unity field starts with `__`) so accidental matches
/// are unlikely.
pub const MARK_PREFIX: &str = "__MARK__";
/// Field separator inside a marker's payload. `␞` (U+241E SYMBOL FOR
/// RECORD SEPARATOR) is printable, JSON.stringify-stable, and outside
/// any plausible user content.
pub const MARK_SEP: char = '\u{241e}';

pub const MARK_TYPE_PPTR: &str = "pptr";
pub const MARK_TYPE_COLOR: &str = "color";

/// Build a `pptr` marker. Empty fields are allowed (a null pptr lands
/// in a map key as the all-empty variant; see the walker).
///
/// `side` is `""` outside of diff dumps; in the diff content endpoint
/// it's `"base"` / `"target"` so the frontend can resolve the marker
/// against the correct manifest (a `-` line's pptr targets the
/// target side, a `+` line's the base side).
pub fn pptr_marker(ref_: &str, target: &str, type_id: &str, file: &str, side: &str) -> String {
    let mut out = String::with_capacity(
        MARK_PREFIX.len() + 7 + ref_.len() + target.len() + type_id.len() + file.len() + side.len(),
    );
    let _ = write!(
        &mut out,
        "{MARK_PREFIX}{MARK_TYPE_PPTR}{MARK_SEP}{ref_}{MARK_SEP}{target}{MARK_SEP}{type_id}{MARK_SEP}{file}{MARK_SEP}{side}",
    );
    out
}

/// Build a `color` marker. `hex` should already include the leading
/// `#` and be exactly 8 lower-case hex chars (`#rrggbbaa`); the
/// frontend regex is uniform on that shape.
pub fn color_marker(hex: &str) -> String {
    format!("{MARK_PREFIX}{MARK_TYPE_COLOR}{MARK_SEP}{hex}")
}

/// Detect a PPtr-shaped map (`{m_FileID: i32, m_PathID: i64}`).
pub fn pptr_from_map(map: &BTreeMap<Value, Value>) -> Option<PPtr> {
    if map.len() != 2 {
        return None;
    }
    let file_v = map.get(&Value::String("m_FileID".to_string()))?;
    let path_v = map.get(&Value::String("m_PathID".to_string()))?;
    Some(PPtr::new(as_file_id(file_v)?, as_path_id(path_v)?))
}

/// Detect a color-shaped map (`{r,g,b,a}` of floats) and return the
/// `#rrggbbaa` payload. HDR components above 1.0 clamp to `ff` — the
/// swatch is a visual hint, not a faithful reproduction.
pub fn color_hex_from_map(map: &BTreeMap<Value, Value>) -> Option<String> {
    if map.len() != 4 {
        return None;
    }
    let r = as_unit_float(map.get(&Value::String("r".to_string()))?)?;
    let g = as_unit_float(map.get(&Value::String("g".to_string()))?)?;
    let b = as_unit_float(map.get(&Value::String("b".to_string()))?)?;
    let a = as_unit_float(map.get(&Value::String("a".to_string()))?)?;
    let to_u8 = |v: f64| (v.clamp(0.0, 1.0) * 255.0).round() as u8;
    Some(format!(
        "#{:02x}{:02x}{:02x}{:02x}",
        to_u8(r),
        to_u8(g),
        to_u8(b),
        to_u8(a),
    ))
}

fn as_unit_float(v: &Value) -> Option<f64> {
    match v {
        Value::F32(x) => Some(*x as f64),
        Value::F64(x) => Some(*x),
        _ => None,
    }
}

/// PPtr fields in the typetree are `int m_FileID; SInt64 m_PathID;` so
/// the deserialiser always emits I32 + I64 — no need to handle the
/// other integer widths.
pub fn as_file_id(v: &Value) -> Option<FileId> {
    match v {
        Value::I32(x) => Some(FileId::new(*x)),
        _ => None,
    }
}

pub fn as_path_id(v: &Value) -> Option<PathId> {
    match v {
        Value::I64(x) => Some(*x),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn s(s: &str) -> Value {
        Value::String(s.to_string())
    }

    #[test]
    fn pptr_marker_round_trip() {
        let m = pptr_marker("obj:5", "Player/Camera", "Camera", "depot.dll", "base");
        // Wire shape: prefix + "pptr" + 5 sep-delimited fields.
        assert_eq!(
            m,
            format!(
                "{MARK_PREFIX}pptr{MARK_SEP}obj:5{MARK_SEP}Player/Camera{MARK_SEP}Camera{MARK_SEP}depot.dll{MARK_SEP}base",
            )
        );
    }

    #[test]
    fn color_marker_uses_hex_payload() {
        assert_eq!(
            color_marker("#ff8800ff"),
            format!("{MARK_PREFIX}color{MARK_SEP}#ff8800ff"),
        );
    }

    #[test]
    fn pptr_from_map_detects_canonical_shape() {
        let mut map = BTreeMap::new();
        map.insert(s("m_FileID"), Value::I32(0));
        map.insert(s("m_PathID"), Value::I64(42));
        let pptr = pptr_from_map(&map).unwrap();
        assert_eq!(pptr.m_PathID, 42);
    }

    #[test]
    fn pptr_from_map_rejects_other_shapes() {
        let mut map = BTreeMap::new();
        map.insert(s("m_FileID"), Value::I32(0));
        // Missing m_PathID.
        assert!(pptr_from_map(&map).is_none());

        // Wrong types for m_FileID / m_PathID.
        let mut map = BTreeMap::new();
        map.insert(s("m_FileID"), Value::I64(0));
        map.insert(s("m_PathID"), Value::I64(1));
        assert!(pptr_from_map(&map).is_none());
    }

    #[test]
    fn color_hex_from_map_canonical_and_clamping() {
        // 1.0 → ff per channel; 0.0 → 00; mid → 80-ish.
        let mut map = BTreeMap::new();
        map.insert(s("r"), Value::F32(1.0));
        map.insert(s("g"), Value::F32(0.0));
        map.insert(s("b"), Value::F32(0.5));
        map.insert(s("a"), Value::F32(1.0));
        assert_eq!(color_hex_from_map(&map).as_deref(), Some("#ff0080ff"));

        // HDR > 1.0 clamps to ff (visual hint, not faithful).
        let mut map = BTreeMap::new();
        map.insert(s("r"), Value::F32(3.5));
        map.insert(s("g"), Value::F32(1.0));
        map.insert(s("b"), Value::F32(1.0));
        map.insert(s("a"), Value::F32(1.0));
        assert_eq!(color_hex_from_map(&map).as_deref(), Some("#ffffffff"));
    }

    #[test]
    fn color_hex_from_map_rejects_wrong_shape() {
        // Three components — not a color.
        let mut map = BTreeMap::new();
        map.insert(s("r"), Value::F32(1.0));
        map.insert(s("g"), Value::F32(0.0));
        map.insert(s("b"), Value::F32(0.0));
        assert!(color_hex_from_map(&map).is_none());
    }
}
